src/Bubble.vala

```diff
@@ -1,103 +1,73 @@
-            case GLib.NotificationPriority.HIGH:
-            case GLib.NotificationPriority.URGENT:
-                content_area.get_style_context ().add_class ("urgent");
-                break;
-            default:
+            if (value.priority >= NotificationPriority.HIGH) {
+                contents.get_style_context ().add_class ("urgent");
+            } else {
                 start_timeout (4000);
-                break;
-        }
-
-        bool default_action = false;
-        bool has_actions = notification.actions.length > 0;
-
-        for (int i = 0; i < notification.actions.length; i += 2) {
-            if (notification.actions[i] == "default") {
-                default_action = true;
-                break;
             }
+
+            content_area.add (contents);
+            content_area.visible_child = contents;
         }
+    }
 
-        contents.action_invoked.connect ((action_key) => {
-            action_invoked (action_key);
-            dismiss ();
-        });
+    private Notification _notification;
+    private bool _has_default;
 
-        button_release_event.connect ((event) => {
-            if (default_action) {
-                action_invoked ("default");
-                dismiss ();
-            } else if (notification.app_info != null && !has_actions) {
+    public Bubble (Notification notification) {
+        Object (notification: notification);
+        action_invoked.connect (dismiss);
+    }
+
+    protected override bool button_release_event (Gdk.EventButton event) {
+        if (_has_default) {
+            action_invoked ("default");
+        } else if (notification.app_info != null) {
+            notification.app_info.launch_uris_async.begin (null, null, null, (obj, res) => {
                 try {
-                    notification.app_info.launch (null, null);
+                    ((AppInfo) obj).launch_uris_async.end (res);
                     dismiss ();
                 } catch (Error e) {
                     critical ("Unable to launch app: %s", e.message);
                 }
-            }
-
-            return Gdk.EVENT_STOP;
-        });
+            });
+        }
 
-        leave_notify_event.connect (() => {
-            if (notification.priority == GLib.NotificationPriority.HIGH || notification.priority == GLib.NotificationPriority.URGENT) {
-                return Gdk.EVENT_PROPAGATE;
-            }
-            start_timeout (4000);
-        });
+        return Gdk.EVENT_STOP;
     }
 
-    public void replace (Notifications.Notification new_notification) {
-        start_timeout (4000);
-
-        var new_contents = new Contents (new_notification);
-        new_contents.show_all ();
-
-        new_contents.action_invoked.connect ((action_key) => {
-            action_invoked (action_key);
-            dismiss ();
-        });
+    protected override bool leave_notify_event (Gdk.EventCrossing event) {
+        if (notification.priority < GLib.NotificationPriority.HIGH) {
```

Comment:
Wasn't there a change recently with notifications because we treat high priority as a normal priority? I would think this is supposed to be less than urgent is that is the case.

Reply:
That's #189, i prefer to keep PRs the most self-contained possible, so that they can be reviewed and merged in any order.